### PR TITLE
encrypted metafeed API (AGAIN)

### PR DESCRIPTION
**Pull request by mixmix**
Cases this has to cover:
1. :x: `recp = sbot.id`
   - this is technically an "own_secret" case, but we will hit bumps because ssb-box2 doesn't know about meta-feeds, and "own" now being many keys
       - on encrption, I _think_ box2 will see this as own-encrypt?
       - on decrypt, box2 will almost certainly NOT try own keys

2. :heavy_check_mark:  `recps = groupId`
   - this case should be fine... as we try group keys on all our messages

3. :x: `recp = someFeedId`
   - this is the DM case, so the shared key should be derived from shardFeedId (author) + someFeedId
   - this should "just work"

4. :x: `recp = POBoxId`

:x: = this PR doesn't handle, and has test cases showing they explicitly callback with an error

### CODE REVIEW

File: api.js [hunk 1]
```diff
@@ -16,19 +17,20 @@ function detailsToVisit(details) {
...
-      deepEqual(feed.metadata, details.metadata || {})
+      (details.metadata ? deepEqual(feed.metadata, details.metadata) : true) &&
+      (details.recps ? deepEqual(feed.recps, details.recps) : true)
     )
   }
 }
 
 exports.init = function (sbot, config) {
   function filter(metafeed, visit, maybeCB) {
+    const cb = maybeCB
```
Comment (mixmix) [inline api.js:28]: I don't know why this code was written like this. I moved this.... either maybeCB isn't there and it errors, or... if I leave the code it definitely errors? Slight improvement

File: package.json [hunk 2]
```diff
@@ -25,8 +25,7 @@
     "pull-defer": "^0.2.3",
     "pull-notify": "^0.1.1",
     "pull-stream": "^3.6.14",
-    "ssb-bfe": "^3.6.0",
-    "ssb-db2": ">=3.0.0 <=6",
+    "ssb-bfe": "ssbc/ssb-bfe#cloaked",
```
Comment (mixmix) [inline package.json:28]: NOTE - depends on this being merged

File: api.js [hunk 3]
```diff
@@ -43,12 +45,12 @@ exports.init = function (sbot, config) {
   }
 
   function find(metafeed, visit, maybeCB) {
+    const cb = maybeCB
```
Comment (staltz) [inline api.js:48]: Yeah it seems the logic for cb here is all upside down. It's supposed to be like `create()` (line 70) because the arguments are flexible and we don't know if we have a callback. I think `find` and `filter` used to be public APIs but now that they're private, we can perhaps assume that the arguments are non-flexible (i.e. cb is always given, and always in the 3rd slot).

Note however that this is completely unrelated to your PR. Don't work on refactoring this in this PR.

File: query.js [hunk 4]
```diff
@@ -61,16 +62,8 @@ exports.init = function (sbot, config) {
...
-        'subfeed',
-        'nonce',
-        'metafeed',
-        'tangles',
-        'type',
-      ]
-      for (const key of Object.keys(content)) {
-        if (ignored.includes(key)) continue
+      for (const key in content) {
+        if (NOT_METADATA.has(key)) continue
```
Comment (staltz) [inline query.js:66]: I think your PR #89 is going to grow more, so we could probably take these changes which are unrelated (they're left over from the reserved-keywords PR) and make it its own PR.

File: test/api/advanced/find-or-create.test.js [hunk 5]
```diff
@@ -151,3 +151,81 @@ test('advanced.findOrCreate (protected metadata fields)', (t) => {
...
+
+          t.deepEqual(msgs[0].value.content.recps, [groupId], 'metafeed/add/derived has recps')
+          // t.deepEqual(typeof msgs[0].value?.meta?.originalContent, 'string', 'metafeed/add/derived is encrypted')
+
+          cb(null)
+        }))
+      }
+    ]
+
+    waterfall(series, cb)
```
Comment (staltz) [inline test/api/advanced/find-or-create.test.js:197]: Could we not add this dependency? I want the tests to run only "raw JS" code and *our* library, to rule out mistakes from other libraries and test only our library. If I understand correctly, we can use just a traditional callback pyramid for this purpose. It's not pretty, but everyone understands what's going on, unlike a new dependency.

**Review by staltz — CHANGES_REQUESTED** [review]

File: test/api/advanced/find-or-create.test.js (hunk 5 above)
Comment (mixmix) [reply test/api/advanced/find-or-create.test.js:197]: The pyramid is annoying to maintain IMO.
Agree we should discuss "standard tools" we're using. Will bring this up in next meeting and change this for now

File: api.js (hunk 3 above)
Comment (mixmix) [reply api.js:48]: I had a test fail because of this, so I fixed it.
I would like some tooling that catches broken code like this (my linting/highlighting sees it in my editor)

Comment (mixmix) [reply api.js:48]: This is an example of something I'm not going to revert / split out.
It's not a good use of time and the code was broken as it was

File: query.js (hunk 4 above)
Comment (mixmix) [reply query.js:66]: they are related. I had to add `recps` to this `ignored` list. So it was add recps as a string to this file, or swap this NOT_METADATA in.

I don't intend to change this one

File: test/api/advanced/find-or-create.test.js (hunk 5 above)
Comment (mixmix) [reply test/api/advanced/find-or-create.test.js]: DONE

**Comment by mixmix** [comment]
@staltz have made changes I'm going to, and highlighted the ones I'm not (along with reasons)
From here, can you review and merge.

You'll see I've pushed back on your requests to put some things in other PRs. I'd like you to consider this as an example of a place for you to concede some flexibility. I've done a lot of flexing my working approach to meet you, and it would feel good if we could both practice that.

**Comment by mixmix** [comment]
Oh, NOTE I'm not working on supporting encryption to FeedId at all here. That's for a future PR (maybe, if we need it).
This change gives any user putting a FeedId in a clear error

File: api.js (hunk 3 above)
Comment (staltz) [reply api.js:48]: I wasn't asking you to do a refactor, I was just suggesting about a new refactor, i.e.

> we can perhaps assume that the arguments are non-flexible

but I was advising to not do this new suggestion in this PR.

File: messages.js [hunk 6]
```diff
@@ -20,10 +20,17 @@ const { deriveFeedKeyFromSeed } = require('./keys')
...
+    feedKeys,
+    nonce,
+    feedpurpose,
+    metadata,
+    recps,
+    encryptionFormat
+  ) {
     const content = {
-      type,
+      type: nonce ? 'metafeed/add/derived' : 'metafeed/add/existing',
```
Comment (staltz) [inline messages.js:33]: @mixmix There's no action point in what I'm going to say, I just want to comment on a pattern that I find annoying. This refactor right here is merely stylistic. I wrote this part and I prefer to have the `const type`, and you just ripped that out, probably because it's not your style, without justifying it. Can I ask you to avoid doing these kind of changes that are (1) unjustified, (2) merely code style that conflicts with the original author's preference.

File: test/api/advanced/find-or-create.test.js [hunk 7]
```diff
@@ -151,3 +151,76 @@ test('advanced.findOrCreate (protected metadata fields)', (t) => {
...
+        },
+        (err, f) => {
+          if (err) t.error(err, 'no error')
+
+          t.deepEqual(f.recps, [groupId], 'FeedDetails contains recps')
+
+          sbot.db.query(where(type('metafeed/add/derived')), toCallback((err, msgs) => {
+            if (err) return cb(err)
+
+            if (msgs.length !== 1) t.equal(msgs.length, 1, 'only one metafeed/add/derived')
```
Comment (staltz) [inline test/api/advanced/find-or-create.test.js:186]: I'm confused, isn't there supposed to be 2 `metafeed/add/derived` msgs? 1 for the root mf adding a shard, and 1 for the shard feed adding the group feed.

Comment (staltz) [reply test/api/advanced/find-or-create.test.js:186]: Also, why the if condition?

```suggestion
            t.equal(msgs.length, 1, 'only one metafeed/add/derived')
```

**Review by staltz — CHANGES_REQUESTED** [review]

File: messages.js (hunk 6 above)
Comment (mixmix) [reply messages.js:33]: This is fair critique. I have a "tidy as I go" approach but we don't agree what "tidy" means.
I'll do my best to not do this (I did consciously leave other areas alone)

File: package.json [hunk 8]
```diff
@@ -25,8 +25,7 @@
     "pull-defer": "^0.2.3",
     "pull-notify": "^0.1.1",
     "pull-stream": "^3.6.14",
-    "ssb-bfe": "^3.6.0",
-    "ssb-db2": ">=3.0.0 <=6",
+    "ssb-bfe": "^3.6.1",
```
[the remark this answers is not in the document]
Comment (mixmix) [reply package.json:28]: nice catch. I just did `npm i ssb-db2@latest` after publishing a new version.
Didn't realise this side effect

Committing revert

File: test/api/advanced/find-or-create.test.js (hunk 7 above)
Comment (mixmix) [reply test/api/advanced/find-or-create.test.js:186]: Good question - I think this is a mistake in the tests to only expect 1. I wonder if it's passing because of a race condition around indexing (sigh).

I'm noticing this is quite common (if it's what it is) - to the point that I want a helper function. e.g.

```js
ssb.db.indexesReady(cb)
```

will investigate and report back

**Comment by mixmix** [comment]
@staltz we agreed to work together towards a shared pattern - that I would push hard for micro-PRs (which I think I've been doing fairly well on), and that you would try flexing if there was something that was a minor step off that (I'm measuring by whether it changes the PR much logically, conceptually, thematically). 

Anyway that's for concession, it feels good for momentum.

Also thanks for the close review - you caught some good potential bugs
:heart:

File: test/api/advanced/find-or-create.test.js (hunk 7 above)
Comment (staltz) [reply test/api/advanced/find-or-create.test.js]: For jitdb, everything you query should be up-to-date with the log. But for leveldb indexes, there is `ssb.db.onDrain(nameOfTheLevelIndex, cb)` as a way of waiting until it's indexed

**Comment by mixmix** [comment]
oops I clicked the "Refresh" icon next to your name which was actually "request another review" (I must have some compulsion to click refresh)

File: test/api/advanced/find-or-create.test.js (hunk 7 above)
Comment (mixmix) [reply test/api/advanced/find-or-create.test.js]: OOOOH, the reason there's only 1 `metafeed/add/derived` is because this is testing the lower level (advanced) API which doesn't do sharding
